test/object.civet

```diff
@@ -1285,7 +1285,15 @@ describe "object", ->
       ---
       x.y()?.z.{a,b}
       ---
-      let ref;(ref = x.y()?.z,{a:ref.a,b:ref.b})
+      let ref;({a:(ref = x.y()?.z).a,b:ref.b})
```

Comment:
This may be out of scope for this particular fix but if `x.y()` ends up undefined then `.a` and `.b` on the ref will throw instead of being short circuited.

Also related to that we should decide what to return in the case of the ref not resolving. I think `undefined` is correct but in the previous example with the `?.` directly before the glob we're returing `{nx: undefined, ny: undefined}`

I think the check will end up being something like `((ref = x.y()) != null ? ((ref=ref.z), { a:ref.a, b:ref.b }) : ref`

Sorry it's dense but note the use of the comma operator inside the first ternary branch to manually update the ref. We may need to do it for each time `?.` appears before the glob.

Essentially manually short circuiting every `?.` along the chain.

Reply:
Hmm, good point: this is probably unexpected behavior that `x?.y.{a,b}` crashes when `x` is undefined. The behavior is not changed by this PR, but this PR's compilation does make it potentially more difficult 

One challenge with your compilation is that globs can appear inside object expressions, e.g. `{c, x?.y.{a,b}}`. That's why I switched to setting the `ref` here to get set in the first element instead of outside in a parenthesized expression; the latter wasn't properly mixable. But of course it's possible to combine the two strategies as you say.

A simpler workaround, but with different behavior, would be to compile `x()?.y.z.{a,b}` into `{a: (ref = x()?.y.z)?.a, b: ref?.b}`.  In other words, add a `?` after `ref` if there was any `?` earlier in the chain. But this would return `{a: undefined, b: undefined}` instead of `undefined`. I'm not totally sure which behavior is better... I could see preferring `{a: undefined, b: undefined}` if e.g. doing the following:

```js
{a, b} := object.{a,b}
```

Although, in that case, we could just write `{a, b} := object`. Maybe a better example is `return`ing a projected object? But again it's unclear which is better/more intuitive: returning `undefined` or an object containing `undefined`s... What do you think?

Reply:
I think ultimately we'll need to short circuit at exactly each `?.` The ones outside the glob prevent adding any keys, the ones inside the glob add the key but its value may be undefined.

Checking later in aggregate wouldn't quite be semantically correct. In this case `x?.y` when `x` is defined but `y` is undefined should throw.

Spacing for clarity:
```
{c, x?.y.{a,b}}
---
let ref;
({ 
  c, 
  ...( (ref = x) != null && (    // short circuit at `x?.`
    ref = x.y,    // update ref
    {a: ref.a, b: ref.b}    // object for spread
  ) )
})
```

For the case of projecting I think we'd want to explicitly show that we'd prefer to ignore the missing object rather than to absorb the error. TS will also warn us if the value may potentially be undefined.

```
{a, b} := {...object?.{a,b}}
```

Though a case could be made that the `?.` immediately preceding the glob `{...}` may prefer constructing the projection anyway.

Reply:
Interesting. The `{...glob}` splat perspective is clarifying: if the `glob` is undefined (because of an earlier failure), it doesn't add anything to the object. I guess this is somewhat in line with our recent object comprehensions too.

I'm still not totally sure which interpretation is correct, but I'm coming around to and you're probably right that `undefined` more natural behavior than `{a: undefined, b: undefined}`.

Note that if we want the latter, we could avoid `...` I think. (I do agree that we need to check each possible failure separately though; my `?.` attempt is indeed broken.)

```js
{c, x()?.y.{a,b}}
---
let xref, yref;
({
  c,
  a: (xref = x()) == null ? undefined :
    yref = xref.y,
    yref.a,
  b: xref == null ? undefined : yref.b
})
```